### PR TITLE
NVMe HA automation support

### PR DESCRIPTION
**Pull request by sunilkumarn417**
# Description
- NVMeoF Gateway HA automation support with IO.

Initial code, so happy to improvise with more review comments and solutions.

http://magna002.ceph.redhat.com/cephci-jenkins/rlepaksh_logs/ha_sanity.log

```
2024-04-19 15:58:55,478 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.ceph.ceph.py:1563 - Running command ceph osd pool delete rbd rbd --yes-i-really-really-mean-it on 10.0.209.9 timeout 600
2024-04-19 15:58:57,093 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.ceph.ceph.py:1597 - Command completed successfully
2024-04-19 15:58:57,094 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.tests.rbd.rbd_utils.py:107 - Command execution complete
2024-04-19 15:58:57,097 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.ceph.ceph.py:1563 - Running command podman --version | awk {'print $3'} on 10.0.208.106 timeout 600
2024-04-19 15:58:58,432 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.run.py:1041 - Podman Version 4.6.1
2024-04-19 15:58:58,435 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.ceph.ceph.py:1563 - Running command docker --version | awk {'print $3'} on 10.0.208.106 timeout 600
2024-04-19 15:58:59,843 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.ceph.ceph.py:1563 - Running command ceph --version | awk '{print $3}' on 10.0.209.9 timeout 600
2024-04-19 15:59:00,441 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.run.py:1057 - ceph Version 19.0.0-2511-g12ef028e
2024-04-19 15:59:00,446 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.run.py:892 - Test <module 'test_ceph_nvmeof_high_availability' from '/Users/sunilkumarn/workspace/cephci/tests/nvmeof/test_ceph_nvmeof_high_availability.py'> passed
Test <module 'test_ceph_nvmeof_high_availability' from '/Users/sunilkumarn/workspace/cephci/tests/nvmeof/test_ceph_nvmeof_high_availability.py'> passed
2024-04-19 15:59:00,447 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.run.py:947 -
All test logs located here: /Users/sunilkumarn/logs/HA-Sanity-with-IO-10

All test logs located here: /Users/sunilkumarn/logs/HA-Sanity-with-IO-10

TEST NAME                        TEST DESCRIPTION                                               DURATION                         STATUS                    COMMENTS
Basic E2ETest Ceph NVMEoF GW H   E2E-Test NVMEoF Gateway HA                                     0:06:54.926948                   Pass
```

image

### CODE REVIEW

**Comment by rahullepakshi** [comment]
@sunilkumarn417 can you upload the logs to magna002 and share me the link, I wanted to co-relate it with code

File: suites/reef/nvmeof/tier-1_nvmeof_ha_sanity.yaml [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+              - node6
+              - node7
+            allow_host: "*"
+          - nqn: nqn.2016-06.io.spdk:cnode2
+            serial: 2
+            bdevs:
+            - count: 2
+              size: 15G
+              lb_group: node7
+            listener_port: 5001
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply suites/reef/nvmeof/tier-1_nvmeof_ha_sanity.yaml:105]: Any available port number should work for the listener. In Vmware env 4420 is the default option, not mandate.
As QE we need coverage for both.

File: suites/reef/nvmeof/tier-1_nvmeof_ha_sanity.yaml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            listeners:
+              - node6
+              - node7
+            allow_host: "*"
+        failover: systemctl-node6
+      desc: E2E-Test NVMEoF Gateway HA
+      destroy-cluster: false
+      module: test_ceph_nvmeof_high_availability.py
+      name: Basic E2ETest Ceph NVMEoF GW HA sanity test
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply suites/reef/nvmeof/tier-1_nvmeof_ha_sanity.yaml:114]: This is E2E case.

[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply suites/reef/nvmeof/tier-1_nvmeof_ha_sanity.yaml:114]: If needed for some other test case, a new test module with same workflows can be developed.

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    listeners = get_nodes_by_ids(cluster, nodes)
+    lb_group_ids = {}
+    for node in listeners:
+        nvmegwcli = NVMeGWCLI(node)
+        listener_config = {
+            "args": {
+                "subsystem": config["nqn"],
+                "traddr": node.ip_address,
+                "trsvcid": config["listener_port"],
+                "host-name": nvmegwcli.fetch_gateway_hostname(),
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply tests/nvmeof/test_ceph_nvmeof_high_availability.py:34]: If we do that, we miss valiation of gw cli command `gw info`. So its always better to take it from nvme-cli command output.

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    ceph_cluster = config["ceph_cluster"]
+
+    # Add Subsystem
+    nvmegwcli.subsystem.add(
+        **{
+            "args": {
+                **sub_args,
+                **{
+                    "max-namespaces": config.get("max_ns", 32),
+                    "enable-ha": config.get("enable_ha", False),
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply tests/nvmeof/test_ceph_nvmeof_high_availability.py:54]: Agreed, let me check on this.

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+                    "rbd-create-image": True,
+                    "size": bdev_cfg["size"],
+                },
+            }
+            with parallel() as p:
+                # Create namespace in gateway
+                for num in range(bdev_cfg["count"]):
+                    ns_args = deepcopy(namespace_args)
+                    ns_args["rbd-image"] = f"{name}-image{num}"
+                    if bdev_cfg.get("lb_group"):
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply tests/nvmeof/test_ceph_nvmeof_high_availability.py:89]: Yes, Thats correct. But as part of coverage we need this as well. 
And `bdev_cfg.get("lb_group")` will not get into the block, if not provided. This should all handled by business logic or test case config, and test module here is very generalized which works for both.

File: suites/reef/nvmeof/tier-1_nvmeof_ha_sanity.yaml [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        gw_nodes:
+          - node6
+          - node7
+        subsystems:                             # Configure subsystems with all sub-entities
+          - nqn: nqn.2016-06.io.spdk:cnode1
+            serial: 1
+            bdevs:
+            - count: 2
+              size: 10G
+              lb_group: node6
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply suites/reef/nvmeof/tier-1_nvmeof_ha_sanity.yaml:93]: Same as above.

**Comment by sunilkumarn417** [comment]
> @sunilkumarn417 can you upload the logs to magna002 and share me the link, I wanted to co-relate it with code

@rahullepakshi Cant upload the log here, will share it privately.

File: tests/nvmeof/workflows/ha.py [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+            if w.expired:
+                raise TimeoutError(
+                    f"{service} Stopping NVMeofGW service failed after 60s timeout.."
+                )
+
+        # Check for Active or Optimized path for failed GWs.
+        # Validate that there should be only one active path.
+        total_outage_time = None
+        fail_time = datetime.datetime.now()
```
Comment (rahullepakshi) [inline tests/nvmeof/workflows/ha.py:324]: In a correct way, I think, right fail_time is when IO pauses or for time being I think we need to move this line to just after 

p.spawn(gw.systemctl.stop, service)

**Comment by rahullepakshi** [comment]
@sunilkumarn417  Implementation looks good for the use case of GW daemon failed but GW node is accessible. Further we need to develop for use cases where we cannot even access failed GW node

File: tests/nvmeof/workflows/ha.py (hunk 7 above)
Comment (sunilkumarn417) [reply tests/nvmeof/workflows/ha.py:324]: Considering this next PR. where we will get outage time per gateway.

**Comment by sunilkumarn417** [comment]
> @sunilkumarn417 Implementation looks good for the use case of GW daemon failed but GW node is accessible. Further we need to develop for use cases where we cannot even access failed GW node

Thanks @rahullepakshi .

Yes, failure injection with systemctl/daemon is one of the backlog item.
We shouldn't be using RPC calls to gateway at any case, and reaching out at generalized response from `ceph nvme-gw`. until we have something in nvme-cli.

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        int - 0 when the execution is successful else 1 (for failure).
+
+    Example:
+
+        # Execute the nvmeof GW test
+            - test:
+                name: Ceph NVMeoF deployment
+                desc: Configure NVMEoF gateways and initiators
+                config:
+                    gw_node: node6
```
Comment (manasagowri) [inline tests/nvmeof/test_ceph_nvmeof_high_availability.py:245]: Just a small doc correction here. You have used gw_nodes as the name to parse and also in test suite. So could you please change gw_node to gw_nodes here as well?

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+            cfg = {
+                "no_cluster_state": False,
+                "config": {
+                    "command": "apply",
+                    "service": "nvmeof",
+                    "args": {"placement": {"nodes": [i.hostname for i in gw_nodes]}},
+                    "pos_args": [rbd_pool],
+                },
+            }
+            test_nvmeof.run(ceph_cluster, **cfg)
```
Comment (manasagowri) [inline tests/nvmeof/test_ceph_nvmeof_high_availability.py:289]: So as of now we don't have the option to specify the cpu_mask parameter is it?

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply tests/nvmeof/test_ceph_nvmeof_high_availability.py:54]: Hey @rahullepakshi ,

For now, Its just the support for now, in future it needs to be taken out. 

And if we look at the test config, it dont have `--enbale-ha` arg and `dict_to_string` method ignores the `False` boolean argument.

So this is the current log

```
2024-04-18 15:07:01,507 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.ceph.nvmegw_cli.execute.py:16 - NVMe CLI command : subsystem add
2024-04-18 15:07:01,507 (cephci.test_ceph_nvmeof_high_availability) [INFO] - cephci.ceph.ceph.py:1563 - Running command podman run --quiet --rm quay.io/barakda1/nvmeof-cli:1.2.1  --server-address 10.0.208.84 --server-port 5500 subsystem add  --subsystem nqn.2016-06.io.spdk:cnode1 --max-namespaces 32 on 10.0.208.84 timeout 600
```

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py (hunk 9 above)
Comment (sunilkumarn417) [reply tests/nvmeof/test_ceph_nvmeof_high_availability.py:289]: In downstream, by default cpu_mask is set to `4`, so far any other test case requirement then it has to be developed as per test case.

**Review by rahullepakshi — APPROVED** [review]
LGTM

File: tests/nvmeof/test_ceph_nvmeof_high_availability.py (hunk 8 above)
Comment (sunilkumarn417) [reply tests/nvmeof/test_ceph_nvmeof_high_availability.py]: updated

**Review by HaruChebrolu — APPROVED** [review]
LGTM

**Review by manasagowri — APPROVED** [review]
Approving assuming that the get_io_stats way of IO validation for the failover and failback scenarios will be implemented as part of another PR.

**Review by rahullepakshi — APPROVED** [review]
LGTM

**Review by HaruChebrolu — APPROVED** [review]
LGTM

**Review by Manohar-Murthy — APPROVED** [review]
Approving bsaed on team review.

**Pull request merged**